### ISSUE TITLE
When running dpo_finetuning_example.ipynb training loss is zero starting from the second step

### ISSUE THREAD

**Issue by fairydreaming** [huggingface/smol-course#61]
When performing DPO fine-tuning with this notebook: [dpo_finetuning_example.ipynb](https://github.com/huggingface/smol-course/blob/main/2_preference_alignment/notebooks/dpo_finetuning_example.ipynb) starting from the second step loss value is zero:
![dpo-zero-loss](https://github.com/user-attachments/assets/c3b54d82-ac5c-4f00-b383-459fe44eec51)

 I don't think this is the expected behavior.

I tried the notebook both locally and in Colab and it happens in both environments.

**Comment by fairydreaming** [comment]
I found that the reason for this was using the model in fp16 precision. For some reason it introduced nan values in gradient computations, which manifested as zero loss values.

When I changed it to `torch_dtype=torch.float32` in model creation:

```
# Model to fine-tune
model = AutoModelForCausalLM.from_pretrained(
    pretrained_model_name_or_path=model_name,
    torch_dtype=torch.float32,
).to(device)
```

I got this:
![dpo-nonan](https://github.com/user-attachments/assets/b890fd95-6ebe-40e7-92e9-568dfe7a2d2a)

I'm going to plot the losses next to see if it actually learns anything.

**Comment by fairydreaming** [comment]
OK, here's the plot of the loss function values:
![dpo_loss](https://github.com/user-attachments/assets/4cd723b0-0897-4b8a-9c04-2e6ecfb9a68e)

Doesn't look very good...

**Comment by burtenshaw** [comment]
Thanks for sharing this. Here are a few points, try them out and feel free to share plots again:
- I would try out `wandb` or tensorboard for plotting because it gives all the standard metrics for each trainer.
- we are mainly looking for an upward trend in reward margins. There is a brief examples of that in the [`DPOTrainer`](https://huggingface.co/docs/trl/main/en/dpo_trainer) docs.
- Here is an example of logs from a [successful DPO run](https://wandb.ai/krasul/huggingface/runs/c54lmder?nw=nwuserkrasul)

**Comment by burtenshaw** [comment]
Also, if you open a [SUBMISSION] pr with your changes, including the plots. We can help you out with a review. 🙂

**Comment by fairydreaming** [comment]
>     * we are mainly looking for an upward trend in reward margins. There is a brief examples of that in the [`DPOTrainer`](https://huggingface.co/docs/trl/main/en/dpo_trainer) docs.

That's interesting, why isn't this mentioned in [dpo.md](https://github.com/huggingface/smol-course/blob/main/2_preference_alignment/dpo.md)? IMHO ability to distinguishing the successful fine-tuning run from the failure is like the most fundamental thing we should learn from the course. Meanwhile, the only info that I found in the dpo.md is:

> During training, carefully monitor the loss convergence

So in my case I carefully monitored the loss divergence ;)

**Comment by burtenshaw** [comment]
Thanks for the continued feedback @fairydreaming . We're still ironing out the creases on this module. I'll get to work on a PR to make it clearer.

**Comment by mdagost** [comment]
I'm seeing the same weirdness with `torch.float16`.  I added in the test split for an eval set:

<img width="1335" alt="Screen Shot 2024-12-11 at 5 20 02 PM" src="https://github.com/user-attachments/assets/14083b63-cf81-4b6c-9cec-7a7021ae17d8" />

When I changed to `torch.float32` I see results that make much more sense, including the increasing reward margins:

<img width="1330" alt="Screen Shot 2024-12-11 at 5 30 56 PM" src="https://github.com/user-attachments/assets/65d3e016-07b0-4b17-9cd4-04df28cc9bc8" />
